src/toxcore/dht/ping_array.rs

```diff
@@ -0,0 +1,206 @@
+/*
+    Copyright © 2017 Zetok Zalbavar <zexavexxe@gmail.com>
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*!
+Manage ping_id.
+Generate ping_id on request packet, check ping_id on response packet.
+*/
+
+use std::io::{ErrorKind, Error};
+use std::collections::HashMap;
+use std::time::{Duration, Instant};
+
+use toxcore::crypto_core::*;
+use toxcore::dht::packed_node::*;
+
+/// Ping timeout in seconds
+pub const PING_TIMEOUT: u64 = 5;
+
+/// size of DHT ping arrays.
+pub const DHT_PING_ARRAY_SIZE: u64 = 512;
+
+/// Entry type of PingArray
+#[derive(Clone, Debug)]
+struct PingArrayEntry {
+    data: Vec<PackedNode>,
+    time: Instant,
+}
+
+/// Generate ping_id, check if ping_id is correct or timedout
+#[derive(Clone, Debug)]
+pub struct PingArray {
+    entries: HashMap<u64, PingArrayEntry>,
+    last_added: u64,
+    total_size: u64,
+    timeout: Duration,
+}
+
+impl PingArray {
+    /// Create new PingArray object
+    pub fn new(total_size: u64, timeout: Duration) -> Option<PingArray> {
+        if total_size == 0 {
+            return None;
+        }
+
+        Some(PingArray {
+            entries: HashMap::new(),
+            last_added: 0,
+            total_size,
+            timeout,
+        })
+    }
+
+    // Clear timed out entries
+    fn ping_array_clear_timedout(&mut self) {
+        let timedout_dur = self.timeout;
+        self.entries.retain(|&_ping_id, ref ping_array_entry| ping_array_entry.time.elapsed() <= timedout_dur);
+    }
+
+    /// Add a data to the PingArray and return a new ping_id.
+    pub fn ping_array_add(&mut self, data: Vec<PackedNode>) -> u64 {
+        self.ping_array_clear_timedout();
+        let index = self.last_added % self.total_size;
+
+        let mut ping_id = random_u64();
+        ping_id = ping_id / self.total_size;
+        ping_id = ping_id * self.total_size;
```

Comment:
Does calling `/ self.total_size` and calling `* self.total_size` on the next line do nothing?

Reply:
It is integer, so it is equal ping_id - (remainder of ping_id % self.total_size)

Reply:
let it be `%` instead of sequential calls to `/` and `*`

Reply:
Ok